server/auto_updater/dataset_updater.py

```diff
@@ -0,0 +1,288 @@
+import csv
+from collections import defaultdict, Counter
+import lzma
+from pathlib import Path
+import pickle
+import tarfile
+import urllib.request as request
+
+import lxml.etree as ET
+import numpy as np
+import pandas as pd
+import spacy
+import tqdm
+
+nlp = spacy.load("en_core_web_sm")
+filter_tag_list = [
+    "sc",
+    "italic",
+    "xref",
+    "label",
+    "sub",
+    "sup",
+    "inline-formula",
+    "fig",
+    "disp-formula",
+    "bold",
+    "table-wrap",
+    "table",
+    "thead",
+    "tbody",
+    "caption",
+    "tr",
+    "td",
+]
+parser = ET.XMLParser(encoding="UTF-8", recover=True)
+
+
+def gather_new_papers(
+    current_pmc_file,
+    word_model,
+    temp_dir_filename="new_papers_added_dir.tsv",
+    temp_embed_filename="new_papers_embedding.tsv",
+    temp_token_counter="new_paper_token_count.tsv",
+):
+    """
+    This function is designed to open a network stream to extract papers from
+    PubMed Central Open Access (PMCOA)'s FTP server. The program will scan for all files and grab
+    any papers that are not contained within the current paper dataset.
+
+    Parameters:
+        current_pmc_file - The listing of papers in PMCOA
+        word_model - The word2vec model to generate paper embeddings
+        temp_dir_filename - a temp file contains newly added papers
+        temp_embed_filename - a file containing new paper embeddingss
+        temp_token_counter - a file containing new tokens from processed papers
+    """
+
+    # Grab Current List
+    current_pmc_list_df = pd.read_csv(current_pmc_file, sep="\t")
+    current_pmc_set = set(current_pmc_list_df.file_path.tolist())
+
+    # Grab file list on ftp server
+    pmc_open_access_url = "ftp://ftp.ncbi.nlm.nih.gov/pub/pmc/oa_bulk/"
+    response = request.urlopen(f"{pmc_open_access_url}")
+    files = response.read().decode("utf-8").splitlines()
+    tar_files = [f.split(" ")[-1] for f in files]
+
+    with open(temp_dir_filename, "w") as dir_file, open(
+        temp_embed_filename, "w"
+    ) as embed_file, open(temp_token_counter, "w") as word_file:
+        dir_writer = csv.DictWriter(
+            dir_file, delimiter="\t", fieldnames=["tarfile", "file_path"]
+        )
+        dir_writer.writeheader()
+
+        embed_writer = csv.DictWriter(
+            embed_file,
+            delimiter="\t",
+            fieldnames=["journal", "document"] + [f"feat_{idx}" for idx in range(300)],
+        )
+        embed_writer.writeheader()
+
+        count_writer = csv.DictWriter(
+            word_file, delimiter="\t", fieldnames=["document", "lemma", "count"]
+        )
+        count_writer.writeheader()
+
+        # Cycle through each tar file on the server
+        for tar_file in tqdm.tqdm(tar_files):
+
+            # If not xml files skip
+            if all([suffix != ".xml" for suffix in Path(tar_file).suffixes]):
```

Comment:
The `[]` inside `()` is unnecessary, so this line can be simplified a little bit , like this:
```if all(suffix != ".xml" for suffix in Path(tar_file).suffixes):```

I noticed that sometimes the FTP site includes a `*.tmp` file, for example, this is the list of files that I found this morning:
```
comm_use.0-9A-B.txt.tar.gz
comm_use.A-B.xml.tar.gz
comm_use.C-H.txt.tar.gz
comm_use.C-H.xml.tar.gz
comm_use.I-N.txt.tar.gz
comm_use.I-N.xml.tar.gz
comm_use.I-N.xml.tar.gz.tmp
comm_use.O-Z.txt.tar.gz
comm_use.O-Z.xml.tar.gz
non_comm_use.0-9A-B.txt.tar.gz
non_comm_use.A-B.xml.tar.gz
non_comm_use.C-H.txt.tar.gz
non_comm_use.C-H.xml.tar.gz
non_comm_use.I-N.txt.tar.gz
non_comm_use.I-N.xml.tar.gz
non_comm_use.O-Z.txt.tar.gz
non_comm_use.O-Z.xml.tar.gz
```
Note the extra file `comm_use.I-N.xml.tar.gz.tmp`? It existed in the FTP site for a while, then disappeared. If you check their website: here: https://ftp.ncbi.nlm.nih.gov/pub/pmc/oa_bulk/
You can see that `comm_use.I-N.xml.tar.gz` was updated this morning. So my guess is that this extra file was generated when they updated the website. When updating is done, the tmp file replaces the original one. If my guess is correct, we should ignore "*.tmp" files, because they are incomplete.